apstools/devices/srs570_preamplifier.py

```diff
@@ -21,12 +21,169 @@
+    """Determine the best settle time for a given combination of parameters.
+
+    Parameters can be strings of indexes.
+
+    """
+    # Convert indexes to string values
+    try:
+        gain_value = gain_values[gain_value]
+    except (TypeError, IndexError):
+        pass
+    try:
+        gain_unit = gain_units[gain_unit]
+    except (TypeError, IndexError):
+        pass
+    try:
+        gain_mode = gain_modes[gain_mode]
+    except (TypeError, IndexError):
+        pass
+    # Get calibrated settle time, or None to use the Ophyd default
+    return settling_times.get((gain_value, gain_unit, gain_mode))
+
+
+class GainMixin:
+    """A signal where the settling time depends on the pre-amp gain.
+
+    Used to introduce a specific settle time when setting to account
+    for the amp's R–C relaxation time when changing gain.
+
+    """
+
+    _settle_time = staticmethod(calculate_settle_time)
+
+    def set(self, value, *, timeout=DEFAULT_WRITE_TIMEOUT, settle_time="auto"):
+        """Set the value of the Signal and return a Status object.
+
+        If put completion is used for this EpicsSignal, the status object will
+        complete once EPICS reports the put has completed.
+
+        Otherwise the readback will be polled until equal to the set point (as
+        in `Signal.set`)
+
+        Parameters
+        ----------
+        value : any
+        timeout : float, optional
+            Maximum time to wait.
+        settle_time: float, optional
+            Delay after the set() has completed to indicate completion
+            to the caller. If "auto" (default), a reasonable settle
+            time will be chosen based on the gain mode on the pre-amp.
+
+        Returns
+        -------
+        st : Status
+
+        See Also
+        --------
+        Signal.set
+        EpicsSignal.set
+
+        """
+        # Determine optimal settling time.
+        if settle_time == "auto":
+            signals = [self.parent.sensitivity_value, self.parent.sensitivity_unit, self.parent.gain_mode]
+            args = [value if self is sig else sig.get() for sig in signals]
+            val, unit, mode = args
+            # Resolve string values to indices if provided
+            if val in gain_values:
+                val = gain_values.index(val)
+            if unit in gain_units:
+                unit = gain_units.index(unit)
+            if mode in gain_modes:
+                mode = gain_modes.index(mode)
+            # Low-drift mode uses the same settling times as low-noise mode
+            if mode == "LOW DRIFT":
+                mode = "LOW NOISE"
+            # Calculate settling time
+            _settle_time = self._settle_time(gain_value=val, gain_unit=unit, gain_mode=mode)
+        else:
+            _settle_time = settle_time
+        return super().set(value, timeout=timeout, settle_time=_settle_time)
+
+
+class GainSignal(GainMixin, EpicsSignal):
```

Comment:
Is it necessary to separate the code of GainMixin from GainSignal?  In addition to the use here, will the mixin be used somewhere else?

Reply:
It was originally going to be used in a pcdsdevices MultiDerivedSignal class. I re-factored it to no longer need it that way. Happy to re-combine the mixin with the signal.

Reply:
Better to recombine if you do not see an additional use.